src/sleep.rs

```diff
@@ -0,0 +1,110 @@
+//! Driver for light and deep sleep.
+//!
+//! Currently implemented Timer and UART wakeup sources.
+
+use esp_idf_sys::*;
+use std::time::Duration;
+
+#[derive(Debug)]
+pub enum WakeupSource {
+    Timer { dur: Duration },
+    Uart { uart_num: i32, threshold: i32 },
+    // TODO: add more sources
+}
+
+impl WakeupSource {
+    pub fn apply(&self) -> Result<(), EspError> {
+        match *self {
+            WakeupSource::Timer { dur } => {
+                esp!(unsafe { esp_sleep_enable_timer_wakeup(dur.as_micros() as u64) })?;
+            }
+            WakeupSource::Uart {
+                uart_num,
+                threshold,
+            } => {
+                esp!(unsafe { uart_set_wakeup_threshold(uart_num, threshold) })?;
+                esp!(unsafe { esp_sleep_enable_uart_wakeup(uart_num) })?;
+            }
+        }
+
+        Ok(())
+    }
+}
+
+pub trait Sleep {
+    fn add_wakeup_source(&mut self, wakeup: WakeupSource) -> Result<(), EspError>;
+    fn sleep(&self) -> Result<(), EspError>;
+
+    fn apply_wakeup_sources(&self, wakeup_sources: &[WakeupSource]) -> Result<(), EspError> {
+        // reset wakeup sources
+        esp!(unsafe { esp_sleep_disable_wakeup_source(esp_sleep_source_t_ESP_SLEEP_WAKEUP_ALL) })?;
+
+        for wakeup in wakeup_sources.iter() {
+            wakeup.apply()?;
+        }
+        Ok(())
+    }
+}
+
+#[derive(Default, Debug)]
+pub struct LightSleep {
+    wakeup_sources: Vec<WakeupSource>,
+}
+
+impl LightSleep {
+    pub fn timer(dur: Duration) -> Result<Self, EspError> {
+        // TODO: checks for duplicate wakeup sources
+
+        let mut lsleep = Self::default();
+        lsleep.add_wakeup_source(WakeupSource::Timer { dur })?;
+        Ok(lsleep)
+    }
+}
+
+impl Sleep for LightSleep {
+    fn add_wakeup_source(&mut self, wakeup: WakeupSource) -> Result<(), EspError> {
+        self.wakeup_sources.push(wakeup);
+        Ok(())
+    }
+
+    fn sleep(&self) -> Result<(), EspError> {
+        self.apply_wakeup_sources(&self.wakeup_sources)?;
+
+        esp!(unsafe { esp_light_sleep_start() })?;
+        Ok(())
+    }
+}
+
+#[derive(Default, Debug)]
+pub struct DeepSleep {
+    wakeup_sources: Vec<WakeupSource>,
+}
+
+impl DeepSleep {
+    pub fn timer(dur: Duration) -> Result<Self, EspError> {
+        let mut dsleep = Self::default();
+        dsleep.add_wakeup_source(WakeupSource::Timer { dur })?;
+        Ok(dsleep)
+    }
+}
+
+impl Sleep for DeepSleep {
+    fn add_wakeup_source(&mut self, wakeup: WakeupSource) -> Result<(), EspError> {
+        // TODO: checks for duplicate wakeup sources
+
+        if matches!(wakeup, WakeupSource::Uart { .. }) {
+            return Err(EspError::from_infallible::<ESP_ERR_INVALID_ARG>());
+        }
+        self.wakeup_sources.push(wakeup);
+        Ok(())
+    }
+
+    fn sleep(&self) -> Result<(), EspError> {
```

Comment:
One more thing to add: I think you are misunderstanding how deep sleep is supposed to work, hence you are equating the deep sleep API with the light sleep API:
- In light sleep - once you call `esp_light_sleep_start`, the chip enters light sleep, but if you are awoken from light sleep, the execution of the program continues **right after** the call to `esp_light_sleep_start`
- In deep sleep - once you call `esp_deep_sleep_start`, the chip enters deep sleep which means ALL the state of the program is completely erased. When you are awoken from deep sleep, the execution of the program will start from the very beginning - with the bootloader, and everything, including - from the beginning of your main function. So in a way, the `deep_sleep`. function signature should be `fn deep_sleep(...) -> !` to [signify that this function never returns execution control back to the caller](https://www.reddit.com/r/rust/comments/3j22vx/what_is_the_meaning_of_as_a_return_type/).

Reply:
I also would suggest you create a `deep_sleep` exaqmple. This way you'll be able to test what I'm saying for real.

Reply:
I do understand the difference, thats why my light and deep examples differ, even if they use a "similar" api. 

The light sleep example starts and infinite loop and the sleep is basically just like any other sleep/delay. The wakeup reason is checked after each wakeup.

The deep sleep example however just check reset/wakeup reason, then do a deep sleep at the end of the program, and expect to end up at the start again afterwards. 

I agree that the deep sleep return type might be abit confusing (more on that later), but apart from that im not sure i understand your concern here